File: track/poll.py
"""TRACK stage, step 1 (D2 PM): poll Gmail threads.

For every vendor with a stored gmail_thread_id, fetch messages that are not yet in
`interactions`, insert them with direction='inbound', and on the first inbound message move
Contacted -> Replied with actor=system.  Each new inbound then goes through track.infer.
P1 dormancy: Contacted with no reply for N days -> Dormant (system).
"""
from __future__ import annotations

import email.utils
import json
import logging
import sqlite3
from datetime import datetime, timedelta, timezone
from typing import Any

from core.llm import LLM
from db.state import current_state, transition, utcnow
from outreach.gmail_client import strip_quoted_reply
from track.infer import apply_proposal, propose

log = logging.getLogger("track.poll")
TRACKED = ("Contacted", "Replied", "In Discussion")


def _addr(s: str) -> str:
    return (email.utils.parseaddr(s or "")[1] or s or "").strip().lower()


def thread_for_prompt(conn: sqlite3.Connection, thread_id: str) -> list[dict[str, str]]:
    rows = conn.execute("SELECT direction, sent_at, subject, body_text FROM interactions WHERE gmail_thread_id = ? "
                        "AND direction IN ('outbound', 'inbound') ORDER BY sent_at, interaction_id", (thread_id,)).fetchall()
    return [{"direction": "out" if r["direction"] == "outbound" else "in", "date": r["sent_at"] or "",
             "subject": r["subject"] or "", "body": r["body_text"] or ""} for r in rows]


def poll(conn: sqlite3.Connection, gmail: Any, llm: LLM | None, dormant_days: int = 10,
         now: datetime | None = None) -> dict[str, int]:
    now = now or datetime.now(timezone.utc)
    own = _addr(gmail.profile_email())
    counts = {"threads": 0, "inbound_new": 0, "replied": 0, "auto_applied": 0, "pending": 0, "no_change": 0, "dormant": 0}
    threads = conn.execute(
        "SELECT DISTINCT i.vendor_id, i.gmail_thread_id FROM interactions i JOIN vendors v USING (vendor_id) "
        "WHERE i.direction = 'outbound' AND i.gmail_thread_id IS NOT NULL AND v.status IN (?, ?, ?)", TRACKED).fetchall()
    for t in threads:
        counts["threads"] += 1
        vendor_id, thread_id = t["vendor_id"], t["gmail_thread_id"]
        known = {r[0] for r in conn.execute(
            "SELECT json_extract(payload, '$.gmail_message_id') FROM events WHERE vendor_id = ? AND payload LIKE '%gmail_message_id%'",
            (vendor_id,))}
        known |= {r[0] for r in conn.execute(
            "SELECT json_extract(llm_summary, '$.gmail_message_id') FROM interactions WHERE gmail_thread_id = ? AND direction = 'inbound' "
            "AND json_valid(llm_summary)", (thread_id,))}
        known.discard(None)
        try:
            messages = gmail.list_thread_messages(thread_id)
        except Exception as exc:  # network / auth: keep polling the other threads
            log.error("thread %s: %s", thread_id, exc)
            continue
        for m in messages:
            mid = m["message_id"]
            if mid in known or _addr(m["from"]) == own:
                continue
            body = strip_quoted_reply(m["body_text"])
            with conn:
                cur = conn.execute(
                    "INSERT INTO interactions (vendor_id, gmail_thread_id, direction, sent_at, subject, body_text, llm_summary) "
                    "VALUES (?, ?, 'inbound', ?, ?, ?, ?)",
                    (vendor_id, thread_id, m["date"], m["subject"], body, json.dumps({"gmail_message_id": mid, "from": m["from"]})))
            interaction_id = int(cur.lastrowid)
            known.add(mid)
            counts["inbound_new"] += 1
            status, stage = current_state(vendor_id, conn)
            if status == "Contacted":
                transition(vendor_id, "Replied", "system", reason=f"inbound email from {m['from']}", evidence_ref=interaction_id,
                           payload={"gmail_message_id": mid}, conn=conn)
                counts["replied"] += 1
                status, stage = "Replied", None
            proposal = propose(llm, status, stage, thread_for_prompt(conn, thread_id), body)
            decision = apply_proposal(conn, vendor_id, interaction_id, proposal)
            with conn:  # keep the message id retrievable for dedup on the next poll
                conn.execute("UPDATE interactions SET llm_summary = ? WHERE interaction_id = ?",
                             (json.dumps({"gmail_message_id": mid, "from": m["from"], "summary": proposal.get("summary"),
                                          "proposal_id": decision.get("proposal_id"), "state": decision["state"]}), interaction_id))
            counts[decision["state"] if decision["state"] in counts else "no_change"] += 1
            with conn:
                conn.execute("UPDATE vendors SET next_action = ?, updated_at = ? WHERE vendor_id = ?",
                             ("review_proposal" if decision["state"] == "pending" else "follow_up", utcnow(), vendor_id))
    cutoff = (now - timedelta(days=dormant_days)).isoformat(timespec="seconds")
    for v in conn.execute("SELECT vendor_id FROM vendors WHERE status = 'Contacted'"):
        last_out = conn.execute("SELECT max(sent_at) FROM interactions WHERE vendor_id = ? AND direction = 'outbound'",
                                (v["vendor_id"],)).fetchone()[0]
        if last_out and last_out < cutoff:
            transition(v["vendor_id"], "Dormant", "system", reason=f"no reply in {dormant_days} days", conn=conn)
            with conn:
                conn.execute("UPDATE vendors SET next_action = 'follow_up_or_close', updated_at = ? WHERE vendor_id = ?",
                             (utcnow(), v["vendor_id"]))
            counts["dormant"] += 1
    return counts

File: src/lib/track/poll.ts
/**
 * Inbound tracking (PRD §7 "Gmail poll -> infer"). For every vendor with an
 * outbound Gmail thread: fetch messages not yet stored, insert them as
 * interactions (direction=inbound), move Contacted -> Replied on the first
 * one (actor=system), then run inference and apply or propose (PRD §6).
 */
import { and, desc, eq, inArray, isNotNull } from "drizzle-orm";

import { getDb, nowIso, type Db } from "@/lib/db";
import { insertInteraction } from "@/lib/db/queries";
import { interactions, proposals, vendors, type Vendor } from "@/lib/db/schema";
import { TransitionError, transition } from "@/lib/db/state";
import { listThreadMessages, profile } from "@/lib/outreach/gmail";
import { unknownMustFields } from "@/lib/shared/must-fields";

import { inferStatus, type Inference, type InferenceInput } from "./infer";

const ACTIVE: Vendor["status"][] = ["Contacted", "Replied", "In Discussion"];
const RESPOND_WITHIN_DAYS = 2;

/** Drop quoted history so inference sees only the new text. */
export function stripQuotedReply(text: string): string {
  const lines = text.replace(/\r\n/g, "\n").split("\n");
  const out: string[] = [];
  for (const line of lines) {
    if (/^On .{5,200} wrote:\s*$/.test(line.trim())) break;
    if (/^-{3,}\s*Original Message\s*-{3,}$/i.test(line.trim())) break;
    if (/^_{10,}$/.test(line.trim())) break;
    if (/^From:\s.+$/.test(line.trim()) && out.length > 0) break;
    if (line.trim().startsWith(">")) continue;
    out.push(line);
  }
  return out.join("\n").replace(/\n{3,}/g, "\n\n").trim();
}

export type InboundMessage = {
  thread_id: string;
  /** message time, ISO; also the dedup key inside a thread */
  sent_at: string;
  subject: string | null;
  body_text: string;
  from?: string | null;
};

export type IngestResult = {
  vendor_id: string;
  interaction_id: number;
  replied: boolean;
  inference: Inference | null;
  applied: boolean;
  proposal_id: number | null;
  error?: string;
};

export type IngestOptions = {
  infer?: boolean;
  /** test seam: replaces the LLM call */
  inferFn?: (input: InferenceInput) => Promise<Inference>;
};

/** One inbound message through the whole pipeline; used by the poller and by the simulate route. */
export async function ingestInbound(vendorId: string, message: InboundMessage, opts: IngestOptions = {}, db: Db = getDb()): Promise<IngestResult> {
  const vendor = db.select().from(vendors).where(eq(vendors.vendor_id, vendorId)).get();
  if (!vendor) throw new Error(`vendor ${vendorId} not found`);
  const body = stripQuotedReply(message.body_text) || message.body_text.trim();
  const now = nowIso();

  const { interaction, replied } = db.transaction((tx) => {
    const row = insertInteraction(
      {
        vendor_id: vendorId,
        direction: "inbound",
        gmail_thread_id: message.thread_id,
        sent_at: message.sent_at,
        subject: message.subject,
        body_text: body,
        llm_summary: null,
      },
      tx,
    );
    let repliedNow = false;
    if (vendor.status === "Contacted") {
      transition({ vendorId, toStatus: "Replied", actor: "system", reason: "first inbound reply", evidenceRef: `interaction:${row.interaction_id}` }, tx);
      repliedNow = true;
    }
    tx.update(vendors)
      .set({ next_action: "respond", due_at: new Date(Date.now() + RESPOND_WITHIN_DAYS * 86_400_000).toISOString(), updated_at: now })
      .where(eq(vendors.vendor_id, vendorId))
      .run();
    return { interaction: row, replied: repliedNow };
  });

  const result: IngestResult = { vendor_id: vendorId, interaction_id: interaction.interaction_id, replied, inference: null, applied: false, proposal_id: null };
  if (opts.infer === false) return result;

  const current = db.select().from(vendors).where(eq(vendors.vendor_id, vendorId)).get() as Vendor;
  if (!ACTIVE.includes(current.status) || current.status === "Contacted") return result;

  const lastOutbound = db
    .select()
    .from(interactions)
    .where(and(eq(interactions.vendor_id, vendorId), eq(interactions.direction, "outbound")))
    .orderBy(desc(interactions.interaction_id))
    .get();

  const input: InferenceInput = {
    vendor: { name: current.name, vendor_type: current.vendor_type, status: current.status, stage: current.diligence_stage },
    message: { from: message.from, subject: message.subject, body, date: message.sent_at },
    lastOutbound: lastOutbound ? { subject: lastOutbound.subject, body: lastOutbound.body_text ?? "" } : null,
    unknownMustFields: unknownMustFields(current),
  };
  let inference: Inference;
  try {
    inference = await (opts.inferFn ?? inferStatus)(input);
  } catch (err) {
    result.error = `inference failed: ${err instanceof Error ? err.message : String(err)}`;
    return result;
  }
  result.inference = inference;

  const summary = `${inference.summary} [${inference.action}: ${inference.reason}]`;
  db.update(interactions).set({ llm_summary: summary }).where(eq(interactions.interaction_id, interaction.interaction_id)).run();

  const propose = () => {
    const p = db
      .insert(proposals)
      .values({
        vendor_id: vendorId,
        interaction_id: interaction.interaction_id,
        to_status: inference.to_status,
        to_stage: inference.to_stage,
        confidence: inference.confidence,
        evidence_snippet: inference.evidence_snippet,
      })
      .returning()
      .get();
    result.proposal_id = p.proposal_id;
  };

  if (inference.action === "apply") {
    try {
      transition(
        {
          vendorId,
          toStatus: inference.to_status,
          toStage: inference.to_stage ?? undefined,
          actor: "llm_inference",
          reason: inference.summary,
          confidence: inference.confidence,
          evidenceRef: `interaction:${interaction.interaction_id}`,
          payload: { evidence_snippet: inference.evidence_snippet, model: inference.model, proposal: false },
        },
        db,
      );
      result.applied = true;
    } catch (err) {
      if (err instanceof TransitionError) propose();
      else throw err;
    }
  } else if (inference.action === "propose") {
    propose();
  }
  return result;
}

export type PollResult = {
  threads: number;
  new_inbound: number;
  replied: number;
  applied: number;
  proposals: number;
  errors: string[];
  results: IngestResult[];
};

/** Fetch every active vendor's Gmail thread and ingest messages not yet stored. */
export async function pollThreads(opts: { vendorId?: string; infer?: boolean } = {}, db: Db = getDb()): Promise<PollResult> {
  const out: PollResult = { threads: 0, new_inbound: 0, replied: 0, applied: 0, proposals: 0, errors: [], results: [] };
  const conds = [eq(interactions.direction, "outbound"), isNotNull(interactions.gmail_thread_id)];
  if (opts.vendorId) conds.push(eq(interactions.vendor_id, opts.vendorId));
  const outbound = db.selectDistinct({ vendor_id: interactions.vendor_id, thread_id: interactions.gmail_thread_id }).from(interactions).where(and(...conds)).all();
  const active = new Set(db.select({ vendor_id: vendors.vendor_id }).from(vendors).where(inArray(vendors.status, ACTIVE)).all().map((v) => v.vendor_id));
  const threads = outbound.filter((t): t is { vendor_id: string; thread_id: string } => Boolean(t.thread_id) && active.has(t.vendor_id));
  if (threads.length === 0) return out;

  const me = (await profile()).email.toLowerCase();
  for (const t of threads) {
    out.threads += 1;
    try {
      const messages = await listThreadMessages(t.thread_id);
      const seen = new Set(
        db.select({ sent_at: interactions.sent_at }).from(interactions).where(and(eq(interactions.gmail_thread_id, t.thread_id), eq(interactions.direction, "inbound"))).all().map((r) => r.sent_at),
      );
      const inbound = messages
        .filter((m) => !m.is_sent && !m.from.toLowerCase().includes(me) && !seen.has(m.internal_date))
        .sort((a, b) => a.internal_date.localeCompare(b.internal_date));
      for (const m of inbound) {
        const r = await ingestInbound(t.vendor_id, { thread_id: t.thread_id, sent_at: m.internal_date, subject: m.subject, body_text: m.body_text, from: m.from }, { infer: opts.infer }, db);
        out.results.push(r);
        out.new_inbound += 1;
        if (r.replied) out.replied += 1;
        if (r.applied) out.applied += 1;
        if (r.proposal_id) out.proposals += 1;
        if (r.error) out.errors.push(`${t.vendor_id}: ${r.error}`);
      }
    } catch (err) {
      out.errors.push(`${t.vendor_id} (${t.thread_id}): ${err instanceof Error ? err.message : String(err)}`);
    }
  }
  return out;
}
